### ISSUE TITLE
[Bug][PyPTO][a5sim] Mixed-kernel expansion omits cross-core synchronization after multiple GM stores

### ISSUE THREAD

**Issue by Crystal-wzy** [hw-native-sys/pypto#2745]
### Problem

A single `pl.spmd` task can store several cube matmul results into partitions
of a GM tensor, then consume that tensor in vector operations. PyPTO splits
the task into AIC and AIV functions but omits the dependency between those
stores and the later loads. The AIV consumer can read before the AIC
producer has finished writing.

Expected: the consumer waits for the required GM writes.
Observed: a fixed-input reproduction completes with **1 instead of 513**
in every output element. Moving the consumer into a dependent task makes
every element pass at zero tolerance.

### Environment

| Component | Revision / version |
| --- | --- |
| pypto-lib | `d6f2920046df7a19c28a5079c096a53b7e600219` |
| PyPTO | `8ecc5d19f0ac1062ae6779ba8aca8cee75c22667` |
| simpler | `39ce891dbb3f665e72e99b4a1387b47012fb18bd` + PR #2186 patches listed below |
| pto-isa | `bc1bd8b9544502a352c6cec1776b9ef00a8303f4` |
| PTOAS | `v0.61` |
| PyTorch | `2.6.0+cpu` |
| Python | `3.10.19` |
| CPU compiler | GCC 15.2.1, C++23, `-O2`, without `-ffast-math` |

The two simpler PR #2186 patches are
`14a46178d83ad31c69ae06adacad2113325206a0` and
`a9a149a4033cd49788582e9f109167c148b00579`.
These are CPU simulator tests; no NPU or model weights are required.

### Error case

From the pypto-lib repository root, with the dependencies above installed:

```bash
export PYTHONPATH=".${PYTHONPATH:+:$PYTHONPATH}"
export DEEPSEEK_V4_VARIANT=pro
timeout --kill-after=10s 300s python \
  models/deepseek_v4_pro/decode_sparse_attn_swa.py -p a5sim
timeout --kill-after=10s 300s python \
  models/deepseek_v4_pro/decode_attention_swa.py -p a5sim
```

Recorded validation failures:

| Model | Output | Mismatched elements | Mismatch ratio | Allowed ratio |
| --- | --- | --- | --- | --- |
| `decode_sparse_attn_swa.py` | `attn_out` | 53945/57344 | 94.0726% | 0.5% |
| `decode_attention_swa.py` | `x_out` | 49944/229376 | 21.7739% | 0.8% |

An element is counted as mismatched when it exceeds the original comparator's
elementwise tolerance. Counts can vary with random inputs and scheduling;
the diagnostic controls replayed fixed failing inputs against unchanged
golden data and tolerances. The separate `kv_cache` failure in SWA attention
is covered by `subissue_simpler_a5sim_arch_initialization.md`.

### Minimal reproduction

Save this as `repro_gm_sync.py` in the pypto-lib repository root. It uses
pypto-lib's `golden` validation helper but imports no model code.

Four matmuls write separate 32-row partitions of `tmp`. A later vector
loop reads `tmp` and adds one. The inputs are all ones, so each matmul
element must be 512 and each output must be 513, both exactly representable
in FP32. `--split` moves only the consumer into a task with `deps=[producer]`.

```python
"""Reproduce missing synchronization between repeated cube stores and vector loads."""
import argparse
import torch
import pypto.language as pl
from golden import TensorSpec, run


@pl.jit
def fused(a: pl.Tensor[[128, 512], pl.BF16], b: pl.Tensor[[512, 128], pl.BF16],
          out: pl.Out[pl.Tensor[[128, 128], pl.FP32]]):
    tmp = pl.create_tensor([128, 128], dtype=pl.FP32)
    with pl.spmd(1):
        idx = pl.tile.get_block_idx()
        for block in pl.unroll(4):
            row = block * 32
            value = pl.matmul(a[row:row+32, :], b, out_dtype=pl.FP32)
            tmp[row:row+32, :] = value
        for row in pl.range(0, 128, 16):
            out[row:row+16, :] = pl.add(tmp[row:row+16, :], 1.0)
    return out


@pl.jit
def separated(a: pl.Tensor[[128, 512], pl.BF16], b: pl.Tensor[[512, 128], pl.BF16],
              out: pl.Out[pl.Tensor[[128, 128], pl.FP32]]):
    tmp = pl.create_tensor([128, 128], dtype=pl.FP32)
    with pl.spmd(1) as producer:
        idx = pl.tile.get_block_idx()
        for block in pl.unroll(4):
            row = block * 32
            value = pl.matmul(a[row:row+32, :], b, out_dtype=pl.FP32)
            tmp[row:row+32, :] = value
    with pl.spmd(1, deps=[producer]):
        idx = pl.tile.get_block_idx()
        for row in pl.range(0, 128, 16):
            out[row:row+16, :] = pl.add(tmp[row:row+16, :], 1.0)
    return out


def golden(tensors):
    tensors["out"].fill_(513.0)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--split", action="store_true")
    args = parser.parse_args()
    result = run(fn=separated if args.split else fused, specs=[
        TensorSpec("a", [128,512], torch.bfloat16, init_value=lambda:torch.ones(128,512)),
        TensorSpec("b", [512,128], torch.bfloat16, init_value=lambda:torch.ones(512,128)),
        TensorSpec("out", [128,128], torch.float32),
    ], golden_fn=golden, config={"platform":"a5sim"}, atol=0, rtol=0)
    if result.error:
        print(result.error)
    print(f"passed={result.passed}")
    raise SystemExit(0 if result.passed else 1)
```

Run both modes with the same dependency versions:

```bash
export PYTHONPATH=".${PYTHONPATH:+:$PYTHONPATH}"
python repro_gm_sync.py
echo "rc=$?"
python repro_gm_sync.py --split
echo "rc=$?"
```

Both modes use `atol=rtol=0`. Recorded results:

| Mode | Expected output | Actual output | Validation | Exit code |
| --- | --- | --- | --- | --- |
| Fused task | All 513 | All 1 | 16384/16384 mismatches | 1 |
| Separate task with `deps=[producer]` | All 513 | All 513 | PASS | 0 |

The fused result can vary with scheduling and initial buffer contents;
the defect is the missing ordering, not a guaranteed stale value of 1.
The control retains the same matmuls, GM accesses, addition, ISA headers
and expected values.

### Root cause and ownership

In `src/ir/transforms/expand_mixed_kernel_pass.cpp`,
`CollectGmCrossLaneSyncs` skips a GM tensor origin with multiple stores:

```cpp
if (origin_stores.size() != 1) continue;
```

The reproduction's four partition stores trigger this restriction.
The collector also restricts the structural relationship of producer and
consumer bodies to keep handshake counts consistent. Unsupported patterns
are left unfenced, allowing an incorrect program to be emitted.

The missing dependency is visible before ISA execution: the generated
`.pto` contains the AIC stores and AIV loads without a cross-core handshake.
The C++ set/wait flags present in this reproduction synchronize pipelines
within a core; they do not order these AIC stores against the AIV loads.
This identifies PyPTO mixed-kernel expansion as the fix location.

A scheduling control provides additional evidence. A C++ harness called
the original generated AIC/AIV functions without the simpler scheduler:

| Execution order | Producer's final GM values | Consumer output |
| --- | --- | --- |
| Consumer first, then producer | All 512 | All 1 |
| Two threads, `std::latch` orders producer completion before consumption | All 512 | All 513 |

The calculation functions and ISA implementation were unchanged.
The latch is a diagnostic ordering control, not an NPU fix. In the full
SWA model, the missing dependency is specifically between the PV GM stores
and normalization reads; earlier QK/softmax communication already exists.

### Proposed fix

Extend the GM dependency analysis to collect the producer writes required
by a later consumer, including multiple partitions and loop bodies. Emit
a cross-core signal after those writes complete and a wait before the
consumer reads, using the supported synchronization mechanism and NPU
store-completion ordering.

Handshake counts must match across loops and branches. Simply removing
the unique-store guard and signaling at every store can deadlock. Patterns
that cannot yet be synchronized safely should be rejected or conservatively
split into tasks instead of being emitted without the required dependency.

For immediate model validation, the following verified workaround applies
to `models/deepseek_v4_pro/decode_sparse_attn_swa.py` at the listed revision:

```diff
--- a/models/deepseek_v4_pro/decode_sparse_attn_swa.py
+++ b/models/deepseek_v4_pro/decode_sparse_attn_swa.py
@@ -202,6 +202,8 @@
                     sparse_blk_li[qk_row : qk_row + H_TILE, 0 : 1] = qk_li[qk_r0 : qk_r0 + H_TILE, 0 : 1]
                     sparse_blk_oi[qk_row : qk_row + H_TILE, 0 : HEAD_DIM] = qk_oi[qk_r0 : qk_r0 + H_TILE, 0 : HEAD_DIM]
 
+    with pl.spmd(T, name_hint="normalize_pack", deps=[qk_tid]) as normalize_tid:
+        qk_t = pl.tile.get_block_idx()
         # Sink, normalize, inverse-RoPE and pack, for every head group of this
         # token. SPARSE_BLOCKS is 1 here, so there is no cross-block softmax
         # state to combine and the partials above are already final.
@@ -255,7 +257,7 @@
             with pl.spmd(
                 O_GROUP_TILE * (O_LORA // PROJ_A_MM_N_TILE // PA_NF_TILE),
                 name_hint="proj_a_mm",
-                deps=[qk_tid],
+                deps=[normalize_tid],
             ) as pa_tid:
                 pa_idx = pl.tile.get_block_idx()
                 pa_local_group = pa_idx // (O_LORA // PROJ_A_MM_N_TILE // PA_NF_TILE)
```

It makes normalization wait for QK/PV completion, then makes projection
wait for normalization. It preserves model math; the cost of the additional
task has not been measured.

### Verification and remaining scope

The minimal reproduction passes at zero tolerance after task separation.
The same workaround makes standalone SWA sparse attention pass.

Model snapshots support the same failure location: completed gather,
QK maxima and PV numerators match their references, but the packed tensor's
last **64/128 heads are zero** in the recorded schedule. The softmax
denominator has only a small FP32 difference. Projection matches the
reference computed from the captured packed tensor, locating the failure
in normalization/packing before projection.

The fixed SWA attention input, captured with seed 2, separates this issue
from the simpler architecture-selection issue:

| Diagnostic change | `kv_cache` | `x_out` |
| --- | --- | --- |
| None | FAIL | FAIL |
| Task separation only | FAIL | PASS |
| A5 initialization only | PASS | FAIL |
| Both | PASS | PASS |

The production compiler fix remains to be implemented. Its regression
coverage should include this fused reproduction, multiple partition stores,
loop/branch handshake counts and existing single-producer cases, checking
both correctness and deadlock freedom. Full model regression, performance
validation and A5 hardware validation remain necessary after that fix.
